### ISSUE TITLE
[FEA] Support reduced precision in balanced k-means 

### ISSUE THREAD

**Issue by cjnolet** [rapidsai/raft#1892]
We should be able to accept reduced-precision inputs and keep the reduced precision throughout the whole computation.

This completes a part of https://github.com/rapidsai/raft/issues/1675.

**Comment by tfeher** [comment]
The k-means clustering implementation calls predict multiple times during its iteration.

The [FusedL2NN](https://github.com/rapidsai/raft/blob/637019e676831c3d38fde4c6e3dfbcae066e1c23/cpp/include/raft/cluster/detail/kmeans_balanced.cuh#L117) operation is used in predict, which under the hood would call [CUTLASS](https://github.com/rapidsai/raft/blob/branch-23.12/cpp/include/raft/distance/detail/fused_l2_nn.cuh#L347) to enable 3xTF32 computations. This way we utilize tensor cores on Ampere (or newer) GPU architectures, and still keep high precision trough the computation. 

Probably we should have three separate issues for the following topics:
- For the ANN workload, it is probably sufficient to use TF32, and this would speed up the clustering. In theory it is a [simple change](https://github.com/rapidsai/raft/pull/1493/files#diff-95028e287ca64a354314a8a6c283313bdf1930f1294dd8dafa2d991363dced66R75-R77), but in practice one might need to tune the GEMM config to reach best perf.
- Furthermore, we plan to support half precision datatsets. For clustering these, FP16 should be preferred.
- Support INT8 efficiently.

**Comment by vinaydes** [comment]
> For the ANN workload, it is probably sufficient to use TF32, and this would speed up the clustering. In theory it is a [simple change](https://github.com/rapidsai/raft/pull/1493/files#diff-95028e287ca64a354314a8a6c283313bdf1930f1294dd8dafa2d991363dced66R75-R77), but in practice one might need to tune the GEMM config to reach best perf.

@tfeher If we find that 1xTF32 path is enough in terms of accuracy and is better performing, we will keep 1xTF32 as a default path for FP32 ANN instead of current 3xTF32. Of course one could override default path if more accuracy is desired. When we implement support for FP16, it will neither use 1xTF32 or 3xTF32. It will probably be FP16 multiplication followed by FP32 accumulation. Is my understanding correct?

**Comment by tfeher** [comment]
>If we find that 1xTF32 path is enough in terms of accuracy and is better performing, we will keep 1xTF32 as a default path for FP32 ANN instead of current 3xTF32.

Yes. I would recommend to add a new option to [index_params](https://github.com/rapidsai/raft/blob/3a7f33f57e35c3076c0f1527c3310ffcd66a866c/cpp/include/raft/neighbors/ivf_pq_types.hpp#L48) that would control the precision used for ANN, and that could have default value as TF32 (i.e 1xTF32). The global default for cutlass GEMMS we would still keep as FP32 (i.e. 3xTF32).

> When we implement support for FP16, it will neither use 1xTF32 or 3xTF32. It will probably be FP16 multiplication followed by FP32 accumulation. Is my understanding correct?

Yes

**Comment by vinaydes** [comment]
@tfeher Thanks for confirming.
